rfc/003-ed25519-verification.md

```diff
@@ -0,0 +1,60 @@
+# RFC 003: Ed25519 Verification
+
+## Changelog
+
+- August 21, 2020: initialized
+
+## Author(s)
+
+- Marko (@marbar3778)
+
+## Context
+
+Ed25519 keys are the only supported key types for Tendermint validators currently. Tendermint-Go wraps the ed25519 key implementation from the go standard library. As more clients are implemented to communicate with the canonical Tendermint implementation (Tendermint-Go) different implementations of ed25519 will be used. Due to [RFC 8032](https://www.rfc-editor.org/rfc/rfc8032.html) not guaranteeing implementation compatibility, Tendermint clients must to come to an agreement of how to guarantee implementation compatibility. [Zcash](https://z.cash/) has multiple implementations of their client and have identified this as a problem as well. The team at Zcash has made a proposal to address this issue, [Zcash improvement proposal 215](https://zips.z.cash/zip-0215).
+
+## Proposal
+
+- Tendermint-Go would adopt [hdevalence/ed25519consensus](https://github.com/hdevalence/ed25519consensus).
+    - This library is implemented as an extension of the go standard library.
+    -
+- Tendermint-rs would adopt [ed25519-zebra](https://github.com/ZcashFoundation/ed25519-zebra)
+    - related [issue](https://github.com/informalsystems/tendermint-rs/issues/355)
+
+As signature verification is one of the major bottlenecks of Tendermint-go, if ZIP 215 is adopted batch verification of signatures will be safe in consensus critical areas.
+
+The change has been recommended to be done in a major release. Although it is not a breaking change if this change were to be rolled out in a minor release a node could construct signatures which could only be verified by nodes which have received updates to the verification function resulting in a possible network halt.
```

Comment:
I think this description is not quite right: in terms of consensus compatibility, this *is* a breaking change, even though in practice it's likely to have no observable effect, because (1) it's backwards compatible with all existing signatures and (2) it's incompatible only on specially-crafted cases that probably won't be part of the chain either before or after the upgrade.

However, because (as you mention here) a node *could* submit those specially-crafted cases, it does still have to be treated as a breaking change.